### PR TITLE
codeintel: Refactor SQLite layer, introduce migration mechanism

### PR DESCRIPTION
**Pull request by efritz**
This work is necessary to change our SQLite db schema for:

- tracking additional information (e.g. code insights)
- applying better encodings without breaking all existing uploads
- changing datastructures within the db for efficiency

This PR adds migrations, but does not do so in a way that affects the current schemas (we only add a version table). 

Please review by commit:

- (1) Add a bundle metadata type instead of passing around raw int/string values in Read/WriteMeta functions. We also drop the lsif/sourcegraph versions here as they're not read anywhere at this time and keep only the numResultChunks value. We may use this type to easily include additional metadata in the future.

- (2) Introduce a *Store struct that backs the SQLite reader and writer. This really just reflects transactions and scanning in the same way as the codeintel/db layer for consistency.

- (3) Remove the writer Flush method and instead have each Write method flush its own inserter. This cleans up the writer interface a bit.

- (4) Add migrations. These are applied in the reader when a database is first opened. The mechanism here is hand-rolled instead of using a third-party tool mainly because the migrations we will be performing in the future are not possible to do only in SQL: we'll want to read rows from the database to re-encode them, which requires us to be able to run arbitrary Go code. We also do not care about migrating databases backwards, so the code introduced here is fairly simple.

The remaining commits are test coverage and cleanup commits.

Note that migration does have test coverage due to it being run in the reader from an old db file, but doesn't have any explicit tests.

### CODE REVIEW

File: internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+// the current (most recent) version of the SQLite writer class. If a serializer output changes in a
+// significant way, it may be necessary to inline the serializer behavior as new migrations are
+// introduced.
+type MigrationFunc func(ctx context.Context, s *store.Store, serializer serialization.Serializer) error
+
+var migrations = []struct {
+	Version       string
+	MigrationFunc MigrationFunc
+}{
+	{"v00000", v0.Migrate},
```
Comment (gbrik) [inline internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go:29]: suggestion:
could these be integers?

File: internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+			continue
+		}
+		if !found {
+			continue
+		}
+
+		if err := runMigration(ctx, s, serializer, migration.Version, migration.MigrationFunc); err != nil {
+			return err
+		}
+	}
```
Comment (gbrik) [inline internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go:59]: nit:
this is hard to follow, i'd rather have a loop to find the current version, and then another to run the migrations afterwards

File: internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	return nil
+}
+
+// getVersion returns the current schema version of the store. If there is no schema_version table,
+// then we assume this database was created prior to this migration mechanism and return the lowest
+// migration version and apply all migrations
+func getVersion(ctx context.Context, s *store.Store) (string, error) {
+	version, exists, err := store.ScanFirstString(s.Query(ctx, sqlf.Sprintf("SELECT version FROM schema_version LIMIT 1")))
+	if err != nil {
+		// TODO(efritz) - is there a better way to match this error?
```
Comment (gbrik) [inline internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go]: suggestion:
could we do a separate query for whether the table exists and not rely on error values at all?

File: internal/codeintel/bundles/persistence/sqlite/store/scan.go [hunk 4]
```diff
@@ -77,3 +77,41 @@ func ScanFirstBytes(rows *sql.Rows, err error) ([]byte, bool, error) {
...
+
+// ScanString populates an integer value from the given scanner.
+func ScanString(scanner scanner) (value string, err error) {
+	err = scanner.Scan(&value)
+	return value, err
+}
+
+// ScanStrings reads the given set of `(int)` rows and returns a slice of resulting values.
+// This method should be called directly with the return value of `*store.Query`.
+func ScanStrings(rows *sql.Rows, err error) ([]string, error) {
```
Comment (gbrik) [inline internal/codeintel/bundles/persistence/sqlite/store/scan.go]: typo:
these comments refer to ints not strings

Comment (efritz) [reply internal/codeintel/bundles/persistence/sqlite/store/scan.go]: Updated.

File: internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go (hunk 3 above)
Comment (efritz) [reply internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go]: Like a query into the schema?

Comment (efritz) [reply internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go]: Updated.

File: internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go (hunk 1 above)
Comment (efritz) [reply internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go:28]: Updating.

File: internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go (hunk 2 above)
Comment (efritz) [reply internal/codeintel/bundles/persistence/sqlite/migrate/migrate.go]: This was simplified after changing versions to ints.
